domain/resource/state/resource.go

```diff
@@ -407,173 +419,568 @@ WHERE uuid = $resourceIdentity.uuid`,
 		}
 
-		// Check resource and unit exists.
-		err = tx.Query(ctx, checkValidResourceStmt, resourceUUID).Get(&resourceUUID)
-		if errors.Is(err, sqlair.ErrNoRows) {
-			return errors.Errorf("resource %s: %w", resourceUUID.UUID, resourceerrors.ResourceNotFound)
-		}
+		// Check that this resource exists and is a container image resource.
+		// Only container image resources can be set as kubernetes application
+		// resource.
+		resourceType, err := st.getResourceType(ctx, tx, resourceUUID)
 		if err != nil {
 			return errors.Capture(err)
+		} else if resourceType != charmresource.TypeContainerImage {
+			return errors.Errorf(
+				"applications can only be set with container image resources, this resource has type %s",
+				resourceType.String(),
+			)
 		}
-		err = tx.Query(ctx, checkValidUnitStmt, unitUUID).Get(&unitUUID)
-		if errors.Is(err, sqlair.ErrNoRows) {
-			return errors.Errorf("resource %s: %w", unitUUID.UnitUUID, resourceerrors.UnitNotFound)
+
+		// Update kubernetes application resource table.
+		var outcome sqlair.Outcome
+		err = tx.Query(ctx, insertK8sAppResourceStmt, k8sAppResource).Get(&outcome)
+		if err != nil {
+			return errors.Capture(err)
 		}
+
+		// Validate that a single row was inserted.
+		rows, err := outcome.Result().RowsAffected()
 		if err != nil {
 			return errors.Capture(err)
+		} else if rows != 1 {
+			return errors.Errorf("inserting kubernetes application resource: expected 1 row affected, got %d", rows)
 		}
+		return nil
+	})
+
+	return err
+}
+
+// SetUnitResource sets the resource metadata for a specific unit.
+// Returns [resourceerrors.UnitNotFound] if the unit id doesn't belong to an
+// existing unit.
+// Returns [resourceerrors.ResourceNotFound] if the resource id doesn't belong
+// to an existing resource.
+func (st *State) SetUnitResource(
+	ctx context.Context,
+	resourceUUID coreresource.UUID,
+	unitUUID coreunit.UUID,
+) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	// Prepare statement to check if the unit/resource is not already there.
+	unitResourceInput := unitResource{
+		ResourceUUID: resourceUUID.String(),
+		UnitUUID:     unitUUID.String(),
+		AddedAt:      st.clock.Now(),
+	}
+	checkUnitResourceQuery := `
+SELECT &unitResource.*
+FROM   unit_resource 
+WHERE  unit_resource.resource_uuid = $unitResource.resource_uuid 
+AND    unit_resource.unit_uuid = $unitResource.unit_uuid`
+	checkUnitResourceStmt, err := st.Prepare(checkUnitResourceQuery, unitResourceInput)
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	// Prepare statement to check that UnitUUID is valid UUID.
+	uUUID := unitNameAndUUID{UnitUUID: unitUUID}
+	checkValidUnitQuery := `
+SELECT &unitNameAndUUID.uuid 
+FROM   unit 
+WHERE  uuid = $unitNameAndUUID.uuid`
+	checkValidUnitStmt, err := st.Prepare(checkValidUnitQuery, uUUID)
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	// Prepare statement to insert a new link between unit and resource.
```

Comment:
What happens if you're updating a unit_resource, not inserting it?

Reply:
Thats a good question. This depends on how we are doing resource update. As I understand it, the `resource_uuid` would not change, so this means that the unit are the resource would never need to be unlinked. The `SetResource` method as it is is idempotent, doing nothing when called would be the correct behavior.

However, if I'm wrong about how we are planning to do resource upgrade, and instead a new resource (with a new UUID) is created, then we are going to have to do something special here, and check if the unit is already linked the charm resource via a different UUID.

For example, if the user runs `attach-resource` and a new resource is uploaded, would we create a new resource UUID for that? Or would we just change the origin, revision, and everything, and remove the old storage.

Reply:
When a new version of the resource is available, the resource_uuid should change. The current is always in application_resource. However the old one can be referred to in unit_resource until the unit actually starts to use the new one.

Once all units or applications stop referencing the resource, it can be deleted from the db and as a blob.